[DOC-3661][DOC-3663] Add a concept for CMEK (#13963)
cockroachdb/docs: v22.1/security-reference/encryption.md

## Encryption at rest

> Industry standard encryption-at-rest is provided at the infrastructure level by your chosen infrastructure-as-a-service (IAAS) provider, either Google Cloud Platform (GCP) or Amazon Web Services (AWS). This helps to protect cloud resources from authorized access over the Information-As-A-Service (IAAS) platform's infrastructure. See documentation for [GCP persistent disk encryption](https://cloud.google.com/compute/docs/disks#pd_encryption) or [AWS Elastic Block Storage](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSEncryption.html).

Reviewer: ```suggestion
Industry standard encryption-at-rest is provided at the infrastructure level by your chosen cloud provider, either Google Cloud Platform (GCP) or Amazon Web Services (AWS). This helps to protect cloud resources from authorized access over the cloud provider's infrastructure. See documentation for [GCP persistent disk encryption](https://cloud.google.com/compute/docs/disks#pd_encryption) or [AWS Elastic Block Storage](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSEncryption.html).
```

Author: I'm going to leave the terminology as it was before I edited the topic. This is under discussion by the docs team and we will be sure to update terminology to keep it consistent. See https://cockroachlabs.atlassian.net/browse/DOC-4584.

In addition to encryption for cluster resources at the infrastructure level, {{ site.data.products.db }} and {{ site.data.products.core }} each include additional optional safeguards for data at rest in clusters.

### CMEK

> For {{ site.data.products.dedicated }} clusters, Customer-Managed Encryption Keys (CMEK) give you more control over how a {{ site.data.products.dedicated }} cluster's data is protected at rest on the cluster's nodes. When CMEK is enabled, your cluster's data is protected by an additional cryptographic key that is entirely within your control, hosted in a supported key-management systems (KMS) platform. Using the KMS's identity access management (IAM) system, you manage CockroachDB's permission to use the key only to encrypt and decrypt. If the key is unavailable, or if CockroachDB no longer has permission to decrypt using the key, the cluster cannot start. To temporarily make the cluster and its data unavailable, such as during a security investigation, you can revoke CockroachDB's access to use the CMEK key or temporarily disable the key within the KMS's infrastructure. To permanently make the cluster's data unavailable, you can delete the CMEK key from the KMS. CockroachDB never has access to the CMEK key materials, and the CMEK key never leaves the KMS.

Reviewer: ```suggestion
For {{ site.data.products.dedicated }} clusters, Customer-Managed Encryption Keys (CMEK) give you more control over how a {{ site.data.products.dedicated }} cluster's data is protected at rest on the cluster's disks. When CMEK is enabled, your cluster's data is protected by a cloud-native KMS cryptographic key that is entirely within your control, hosted in a supported key-management systems (KMS) platform. Using the KMS's identity access management (IAM) system, you manage CockroachDB Dedicated's permission to use the key only to encrypt and decrypt. If the key is unavailable, or if CockroachDB Dedicated no longer has permission to decrypt using the key, the cluster cannot start. To temporarily make the cluster and its data unavailable, such as during a security investigation, you can revoke CockroachDB Dedicated's access to use the CMEK key or temporarily disable the key within the KMS' infrastructure. To permanently make the cluster's data unavailable, you can delete the CMEK key from the KMS. CockroachDB Dedicated never has access to the CMEK key material, and the CMEK key never leaves the KMS.
```

To learn more, see [Customer-Managed Encryption Keys](/docs/cockroachcloud/cmek.html) and [Managing Customer-Managed Encryption Keys (CMEK) for {{ site.data.products.dedicated }}](/docs/cockroachcloud/managing-cmek.html).

When CMEK is enabled, the **Encryption** option appears to be disabled in the [DB Console](../ui-overview.html), because this option refers to [Encryption At Rest (Enterprise)](#encryption-at-rest-enterprise), which is a feature of {{ site.data.products.core }} clusters.

### Encryption at Rest (Enterprise)

> In the context of {{ site.data.products.core }}, customers are responsible for managing their own encryption at rest. Usage of recommended IAAS providers such as GCP and AWS includes the encryption at rest services they offer by default.

Reviewer: ```suggestion
In the context of {{ site.data.products.core }}, customers are responsible for managing their own encryption at rest. Usage of recommended cloud providers such as GCP and AWS includes the encryption at rest services they offer by default.
```

Author: https://cockroachlabs.atlassian.net/browse/DOC-4584

> In addition, the [Encryption at Rest (Enterprise) feature](../enterprise-licensing.html) provides transparent encryption of a node's data on the local disk. It allows encryption of all files on disk using [AES](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) in [counter mode](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#Counter_(CTR)), with all key sizes allowed.

Reviewer: ```suggestion
In addition, the [Encryption at Rest (Enterprise) feature](../enterprise-licensing.html) provides transparent encryption of a data on cluster disks. It allows encryption of all files on disk using [AES](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) in [counter mode](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#Counter_(CTR)), with specific key sizes allowed.
```

Note that backups taken with the [`BACKUP`](../backup.html) statement **are not encrypted** even if Encryption at Rest is enabled. Encryption at Rest only applies to the CockroachDB node's data on the local disk. If you want encrypted backups, you will need to encrypt your backup files using your preferred encryption method.
{{site.data.alerts.end}}

## Encryption keys used by CockroachDB clusters

To allow arbitrary rotation schedules and ensure security of the keys, CockroachDB uses multiple layers of keys:

- **Store key**: A cluster's _store key_ is a _key encryption key (KEK) that CockroachDB uses to encrypt the cluster's data keys (see below).

  For CockroachDB Self-Hosted clusters, you provide the store key and give its location to CockroachDB when starting the cluster. The store key file must contain 32 bytes (the key ID) followed by the key (16, 24, or 32 bytes). The size of the key dictates the version of AES to use (AES-128, AES-192, or AES-256). For an example showing how to create a store key, see [Generating Key Files](/docs/{{site.versions["stable"]}}/encryption.html#generating-store-key-files).

>   For {{ site.data.products.db }} clusters, the store key is created automatically when the cluster is created. For {{ site.data.products.dedicated }} clusters with [Customer-Managed Encryption Keys (CMEK)](/docs/cockroachcloud/cmek.html) enabled, the store key is encrypted with the CMEK key (see below) before being provided to cluster nodes.

Reviewer: No, we don't use encryption-at-rest by default on Dedicated clusters.

Author: I'm confused comparing your comment to the text -- I don't think we assert that it's enabled by default, only when CMEK is enabled. Is it unclear?

  Since very little data is encrypted using this key, it can have a very long lifetime without risk of reuse.

- **Data key**: A cluster's data keys are _data encryption keys_ used to protect your cluster's data on each node's filesystem. Data keys are persisted in a key registry file that is encrypted using the cluster's store key.

  The data key is always the same length as the store key. CockroachDB generates a new data key in the following circumstances:

  - If encryption has just been enabled.
  - If the store key size has changed.
  - When the data key lifetime has expired (by default every 7 days).

  When data is written to the cluster, the current data key is used to encrypt it. When data is read from the cluster, it is decrypted using the data key that was used to encrypt it.

  CockroachDB does not currently force re-encryption of older files but instead relies on normal [storage engine](../architecture/storage-layer.html) churn to slowly rewrite all data with the desired encryption.

  Data keys have short lifetimes to avoid reuse.

> - **CMEK key**: For a {{ site.data.products.dedicated }} cluster with [CMEK](/docs/cockroachcloud/cmek.html) enabled, the CMEK key encrypts the store key at rest on the cluster nodes' filesystems. The CMEK key itself never resides in the cluster and is managed by your organization in a supported cloud-native key-management system (KMS). If the CMEK key is not available or CockroachDB does not have permission to use it, the cluster cannot start and its data is inaccessible.

Reviewer: ```suggestion
- **CMEK key**: For a {{ site.data.products.dedicated }} cluster with [CMEK](/docs/cockroachcloud/cmek.html) enabled, the CMEK key encrypts the store key at rest on the cluster disks. The CMEK key itself never resides in the cluster and is managed by your organization in a supported cloud-native key-management system (KMS). If the CMEK key is not available or CockroachDB does not have permission to use it, the cluster cannot start and its data is inaccessible.
```
